site/en/blog/renderingng-architecture/index.md

```diff
@@ -0,0 +1,508 @@
+
+### The pipeline stages
+
+In the preceeding diagram, stages are notated with colors indicating in which thread or process they execute:
+
+- Render process main thread: green
+- Render process compositor: yellow
+- Viz process: orange
+
+In some cases they can execute in multiple places,
+depending on the circumstance,
+which is why some have two colors.
+
+The stages are:
+
+1. _Style:_ apply CSS to the DOM, and create _computed styles_.
+1. _Animate:_ change computed styles and mutate _property trees_ over time based on declarative timelines.
+1. _Layout:_ determine the size and position of DOM elements on the screen,
+and create the _immutable fragment tree_.
+1. _Pre-paint:_ compute property trees and
+[invalidate](https://en.wikipedia.org/wiki/Cache_invalidation)
+any existing _display lists_ and GPU _texture tiles_ as appropriate.
+1. _Scroll:_ update the scroll offset of documents and scrollable DOM elements, by mutating property trees.
+1. _Paint:_ compute a display list that describes how to raster GPU texture tiles from the DOM.
+1. _Commit:_ copy property trees and the display list to the compositor thread.
+1. _Layerize:_ break up the display list into a _composited layer list_ for independent rasterization and animation.
+1. _Raster, decode and paint worklets:_ turn display lists, encoded images, and paint worklet code, respectively, into
+[GPU texture tiles](https://en.wikipedia.org/wiki/Tiled_rendering).
+1. _Activate:_ create a _compositor frame_ representing how to draw and position GPU tiles to the screen, together with any visual effects.
+1. _Aggregate:_ combine compositor frames from all the visible compositor frames into a single, global compositor frame.
+1. _Draw:_ execute the aggregated compositor frame on the GPU to create pixels on-screen.
+
+Stages of the rendering pipeline can be skipped if they aren't needed.
+For example, animations of visual effects, and scroll, can skip layout, pre-paint and paint.
+This is why animation and scroll are marked with yellow and green dots in the diagram.
+If layout, pre-paint, and paint can be skipped for visual effects,
+they can be run entirely on the compositor thread and skip the main thread.
+
+Browser UI rendering is not depicted directly here,
+but can be thought of as a simplified version of this same pipeline
+(and in fact its implementation shares much of the code).
+Video (also not directly depicted)
+generally renders via independent code that decodes frames into GPU texture tiles
+that are then plugged into compositor frames and the draw step.
+
+## Process and thread structure
+
+{% Aside %}
+There are more CPU processes and threads in the Chromium architecture than pictured here.
+These diagrams focus only on those critical to rendering.
+{% endAside %}
+
+### CPU processes
+
+The use of multiple CPU processes achieves performance and security isolation
+between sites and from browser state,
+and stability and security isolation from GPU hardware.
+
+{% Img
```

Comment:
Consider using some crop of this as a header image, it will make the post stand out when shared on social media.

Reply:
last time the author decided not to go with a header image, @chrishtr let me know what you think.

Reply:
Let's skip it.